dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/doi/DOIBackgroundReservationService.java

```diff
@@ -0,0 +1,111 @@
+import edu.harvard.iq.dataverse.globalid.DOIDataCiteServiceBean;
+import edu.harvard.iq.dataverse.persistence.GlobalId;
+import edu.harvard.iq.dataverse.persistence.dataset.Dataset;
+import edu.harvard.iq.dataverse.persistence.dataset.DatasetRepository;
+import edu.harvard.iq.dataverse.settings.SettingsServiceBean;
+import io.vavr.control.Option;
+import io.vavr.control.Try;
+import org.apache.commons.lang.math.NumberUtils;
+
+import javax.annotation.PostConstruct;
+import javax.ejb.Singleton;
+import javax.ejb.Startup;
+import javax.inject.Inject;
+import java.sql.Timestamp;
+import java.util.Date;
+import java.util.List;
+import java.util.Timer;
+import java.util.TimerTask;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Class designed to reserve doi's in the background.
+ */
+@Startup
+@Singleton
+class DOIBackgroundReservationService {
+
+    private static final Logger logger = Logger.getLogger(DOIBackgroundReservationService.class.getCanonicalName());
+
+    private SettingsServiceBean settingsServiceBean;
+    private DatasetRepository datasetRepository;
+    private DatasetDao datasetDao;
+    private DOIDataCiteServiceBean doiDataCiteService;
+
+    private final Timer timer = new Timer();
+
+    public DOIBackgroundReservationService() {
+    }
+
+    @Inject
+    public DOIBackgroundReservationService(SettingsServiceBean settingsServiceBean, DatasetRepository datasetRepository,
+                                           DatasetDao datasetDao, DOIDataCiteServiceBean doiDataCiteService) {
+        this.settingsServiceBean = settingsServiceBean;
+        this.datasetRepository = datasetRepository;
+        this.datasetDao = datasetDao;
+        this.doiDataCiteService = doiDataCiteService;
+    }
+
+    @PostConstruct
+    void startReservation() {
+        reserveDoiPeriodically(timer);
+    }
+
+    /**
+     * Creates a timer which will reserve doi's in interval provided by 'DoiBackgroundReservationInterval'.
+     * Run method has to be wrapped in Try otherwise Transaction rollback could destroy timer setup.
+     */
+    void reserveDoiPeriodically(Timer timer) {
+        String provider = settingsServiceBean.getValueForKey(SettingsServiceBean.Key.DoiProvider);
+
+        if (provider.equals("DataCite")) {
+
+            Option<Long> intervalInMs = Option
+                    .of(settingsServiceBean.getValueForKey(SettingsServiceBean.Key.DoiBackgroundReservationInterval))
+                    .filter(NumberUtils::isNumber)
+                    .map(Long::parseLong);
+
+            if (intervalInMs.isEmpty() || intervalInMs.get() < 0) {
+                return;
+            }
+
+            timer.schedule(new TimerTask() {
+                               @Override
+                               public void run() {
+                                   Try.run(() -> registerDataCiteIdentifier());
+                               }
+                           }
+                    , 0, intervalInMs.get());
```

Comment:
There is no cleanup code for the timer

Reply:
You mean cleanup if exception occurs or app terminates?

Reply:
When the app terminates.

Reply:
But if the app terminates then the thread will be destroyed so is there any reason to call timer.cancel()?

Reply:
For example when I did deploy the war in glassfish and after that i did undeploy it (no stop-domain command) I've got something like this in logs:
```
[2021-06-09T11:29:02.309+0000] [glassfish 4.1] [WARNING] [] [edu.harvard.iq.dataverse.doi.DOIBackgroundReservationService] [tid: _ThreadID=178 _ThreadName=Timer-8] [timeMillis: 1623238142309] [levelValue: 900] [[
  registerDataCiteIdentifier]]

[2021-06-09T11:29:02.311+0000] [glassfish 4.1] [SEVERE] [] [] [tid: _ThreadID=178 _ThreadName=Thread-9] [timeMillis: 1623238142311] [levelValue: 1000] [[
  Exception in thread "Timer-8"]]

[2021-06-09T11:29:02.312+0000] [glassfish 4.1] [SEVERE] [] [] [tid: _ThreadID=178 _ThreadName=Thread-9] [timeMillis: 1623238142312] [levelValue: 1000] [[
  java.lang.IllegalStateException: This web container has not yet been started
        at org.glassfish.web.loader.WebappClassLoader.loadClass(WebappClassLoader.java:1674)
        at org.glassfish.web.loader.WebappClassLoader.loadClass(WebappClassLoader.java:1633)
        at io.vavr.control.Try.run(Try.java:121)
        at edu.harvard.iq.dataverse.doi.DOIBackgroundReservationService$1.run(DOIBackgroundReservationService.java:81)
        at java.util.TimerThread.mainLoop(Timer.java:555)
        at java.util.TimerThread.run(Timer.java:505)]]
```

Reply:
Hmm, ok. I wasn't able to recreate it but I will just use try{} finally{} then.

Reply:
I was thinking more of something like:
```
@PreDestroy
preDestroy() { timer.cancel(); }
```
Surrounding `.schedule` will not change anything.